src/components/tl/ucp/tl_ucp_team.c

```diff
@@ -86,6 +97,15 @@ UCC_CLASS_INIT_FUNC(ucc_tl_ucp_team_t, ucc_base_context_t *tl_context,
                  "topo is not available, disabling ranks reordering");
         self->cfg.use_reordering = 0;
     }
+    if (self->topo && !IS_SERVICE_TEAM(self) && self->topo->topo->sock_bound) {
+        max_radix       = ucc_min(UCC_TL_TEAM_SIZE(self),
+                                  ucc_topo_min_socket_size(self->topo));
+
+        self->opt_radix = ucc_kn_get_opt_radix(UCC_TL_TEAM_SIZE(self),
+                                               max_radix);
+    }
+    ucc_tl_ucp_switch_auto_radix_in_range(self);
```

Comment:
i think this is bad solution.

I think your change now just breaks the parameter setting option for user. If user sets UCC_TL_UCP_SRA_RADIX=3, then during parsing mranged value would get "default_value=3" and here you just overwrite it to your opt_radix value ignoring the fact that user explictely set it.

What i suggested was: don't modify mrage parameter interface, neither alter its content after user it was parsed.
Instead modify ucc_tl_ucp_get_radix_from_range function. Let take the last argument "default_radix" that it will return for AUTO case, smthe like:
```
static inline unsigned
ucc_tl_ucp_get_radix_from_range(ucc_tl_ucp_team_t *team,
                                size_t             msgsize,
                                ucc_memory_type_t  mem_type,
                                ucc_kn_radix_t       default_radix,
                                ucc_mrange_uint_t *p)
{
    unsigned radix;

    radix = ucc_mrange_uint_get(p, msgsize, mem_type);

    if (UCC_UUNITS_AUTO == radix) {
           return default_radix;
    }
    return radix;
}
```
then in ucc_tl_ucp_allreduce_sra_knomial_frag_init you can do:
```
cfg_radix = ucc_tl_ucp_get_radix_from_range(tl_team,
                                                count * ucc_dt_size(dtype),
                                                mem_type, team->opt_radix, p);
```
same for bcast. later you can have custom opt radixes for different collectives: team->sra_allreduce_opt_radix, team->bcast_sag_opt_radix, etc.

For kn allreduce you can just keep default 4 rthrough parser interface or extend same way as proposed here.

Reply:
My current implementation doesn't break user param setting, since it will update default value only for auto case (also if user sets UCC_TL_UCP_ALLREDUCE_SRA_KN_RADIX=0-64k:3,auto , then it will only use default value for this auto case.

But, I will say that it is more correct to use your suggestion since this way we get this updated value when only coll needs it as apposed to my solution which will set it in team create and then again fetch it during coll. So I will change to your solution.

Reply:
have you verified you statement above?
If add this line:
```
(base) vpetrov@NB-1996-wsl:~/workspace/ucc$ git diff
diff --git a/src/components/tl/ucp/allreduce/allreduce_sra_knomial.c b/src/components/tl/ucp/allreduce/allreduce_sra_knomial.c
index 1e95409..ee4abcb 100644
--- a/src/components/tl/ucp/allreduce/allreduce_sra_knomial.c
+++ b/src/components/tl/ucp/allreduce/allreduce_sra_knomial.c
@@ -119,6 +119,7 @@ static ucc_status_t ucc_tl_ucp_allreduce_sra_knomial_frag_init(
                                                   UCC_TL_TEAM_SIZE(tl_team),
                                                   count);

+printf("starting SRA with radix: %d\n", radix);
     /* 1st step of allreduce: knomial reduce_scatter */
     UCC_CHECK_GOTO(
         ucc_tl_ucp_reduce_scatter_knomial_init_r(&args, team, &task, radix),
```
Then run with current master:
```
(base) vpetrov@NB-1996-wsl:~/workspace/ucc$ mpirun --bind-to none -x xUCC_TL_UCP_ALLREDUCE_SRA_KN_RADIX=2  -np 3 ./install/bin/ucc_test_mpi -c allreduce -I 0 -d float32 -t world -O 0 -m 65536:65536 -o sum

===== UCC MPI TEST INFO =======
   seed        : 12970
===============================

starting SRA with radix: 3
starting SRA with radix: 3
starting SRA with radix: 3

===== UCC MPI TEST REPORT =====
   total tests : 1
   passed      : 1
   skipped     : 0
   failed      : 0
   elapsed     : 0s
(base) vpetrov@NB-1996-wsl:~/workspace/ucc$ mpirun --bind-to none -x UCC_TL_UCP_ALLREDUCE_SRA_KN_RADIX=2  -np 3 ./install/bin/ucc_test_mpi -c allreduce -I 0 -d float32 -t world -O 0 -m 65536:65536 -o sum

===== UCC MPI TEST INFO =======
   seed        : 12976
===============================

starting SRA with radix: 2
starting SRA with radix: 2
starting SRA with radix: 2

===== UCC MPI TEST REPORT =====
   total tests : 1
   passed      : 1
   skipped     : 0
   failed      : 0
   elapsed     : 0s
```
Note how setting RADIX=2 actually sets it.
Now running with your PR:
```
(base) vpetrov@NB-1996-wsl:~/workspace/ucc$ git show HEAD --quiet
commit f14ca7259a9d796601466bd6d65853070638330c (HEAD -> shimmy)
Author: Shimmy Balsam <sbalsam@nvidia.com>
Date:   Tue Jun 13 21:13:43 2023 +0300

    REVIEW: third review fixes
(base) vpetrov@NB-1996-wsl:~/workspace/ucc$ mpirun --bind-to none -x UCC_TL_UCP_ALLREDUCE_SRA_KN_RADIX=2  -np 3 ./install/bin/ucc_test_mpi -c allreduce -I 0 -d float32 -t world -O 0 -m 65536:65536 -o sum

===== UCC MPI TEST INFO =======
   seed        : 13064
===============================

starting SRA with radix: 3
starting SRA with radix: 3
starting SRA with radix: 3

===== UCC MPI TEST REPORT =====
   total tests : 1
   passed      : 1
   skipped     : 0
   failed      : 0
   elapsed     : 0s
```
Note, how setting RADIX=2 is ignored.

I don't see how " it will update default value only for auto case" this is true. You call  ucc_tl_ucp_switch_auto_radix_in_range unconditionally. So mranged param "default" value will get overwritten either to UCC_UUINTS_AUTO_RADIX which is 4, or to your team->opt_radix.

Reply:
switched to your solution